### PR TITLE
Do not load useless columns from the index in memory

### PR DESCRIPTION
**Pull request by alexey-milovidov**
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
If the table's primary key contains mostly useless columns, don't keep them in memory. This is controlled by a new setting `primary_key_ratio_of_unique_prefix_values_to_skip_suffix_columns` with the value `0.9` by default, which means: for a composite primary key, if a column changes its value for at least 0.9 of all the times, the next columns after it will be not loaded.

### CODE REVIEW

**Comment by robot-clickhouse-ci-2** [comment]
*This is an automated comment for commit 31de27b149ab2922647ff7d9141871330cc9d743 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/60255/31de27b149ab2922647ff7d9141871330cc9d743/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/60255/31de27b149ab2922647ff7d9141871330cc9d743/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

File: tests/queries/0_stateless/02998_primary_key_skip_columns.reference [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+15908
+108
+120
+2334
+19
+Key size: 	2400000
+100000
+14954
+798
+15907
```
Comment (alexey-milovidov) [inline tests/queries/0_stateless/02998_primary_key_skip_columns.reference:13]: ?

File: src/Core/Field.h [hunk 2]
```diff
@@ -590,7 +590,7 @@ class Field
 
         switch (which)
         {
-            case Types::Null: return true;
```
Comment (alexey-milovidov) [inline src/Core/Field.h:593]: Here was an error - it incorrectly compared NEGATIVE_INFINITY and POSITIVE_INFINITY.

**Comment by alexey-milovidov** [comment]
> ClickHouse Stateless Tests Flaky Check (asan)

I didn't find any issues other than the query can be slow with ThreadFuzzer.
If it shows up in master, we will disable sanitizer/debug runs for this test.

**Comment by alexey-milovidov** [comment]
> Performance Comparison Aarch64 [1/4] — Failed to parse the report. 

This is an error in our CI, @maxknv

**Comment by Avogar** [comment]
> I didn't find any issues other than the query can be slow with ThreadFuzzer.
If it shows up in master, we will disable sanitizer/debug runs for this test.

Happened in private master:
https://d1k2gkhrlfqv31.cloudfront.net/clickhouse-test-reports-private/0/b89d5164bb72321ba0c7b5abb1bf937d9f851553/stateless_tests__msan__distributed_cache__s3_storage__[3_3].html

**Comment by canhld94** [comment]
@alexey-milovidov this feature can cause bug in `FINAL`. Remind that to parallelize `FINAL` we will split intersect ranges into several non-intersect layers and then process each layer independently. The splitting into layers logic requires exact value of primary keys. I don't have reproduce example now but we're observing it in our prod.
cc @kitaisreal as well.

**Comment by canhld94** [comment]
May be we can only enable this feature for ordinary merge tree.

**Comment by canhld94** [comment]
> @alexey-milovidov this feature can cause bug in `FINAL`. Remind that to parallelize `FINAL` we will split intersect ranges into several non-intersect layers and then process each layer independently. The splitting into layers logic requires exact value of primary keys. I don't have reproduce example now but we're observing it in our prod.
> 
> cc @kitaisreal as well.

To be more specific, it may lead to wrong query result with FINAL

**Comment by alexey-milovidov** [comment]
@canhld94, I remember it was fixed shortly after this pull request.

**Comment by canhld94** [comment]
> @canhld94, I remember it was fixed shortly after this pull request.

Yes, I found that it was fixed in this PR https://github.com/ClickHouse/ClickHouse/pull/62268. Our baseline version is not backported.

Nevertheless, there's a performance tradeoff. It's understandable because if only some prefix of primary keys are used for part splitting, number of non-intersect layers will be reduced.

With #62268
```
Query id: f1ecea2c-c32a-4812-96e6-85783e7329ba

Ok.

0 rows in set. Elapsed: 49.715 sec. Processed 39.82 million rows, 3.22 GB (801.00 thousand rows/s., 64.84 MB/s.)
Peak memory usage: 10.29 GiB.
```

Set `primary_key_ratio_of_unique_prefix_values_to_skip_suffix_columns = 0`:
```
Query id: 089ebf7b-a25b-4581-9881-453730f9b342

Ok.

0 rows in set. Elapsed: 3.248 sec. Processed 40.74 million rows, 3.30 GB (12.54 million rows/s., 1.02 GB/s.)
Peak memory usage: 12.01 GiB.
```

In our case we can set `primary_key_ratio_of_unique_prefix_values_to_skip_suffix_columns = 0` to default, but I think it worth to mention this trade-off somewhere else in the document.

**Pull request merged**